src/transport/MessageEndpointServer.cpp

```diff
@@ -0,0 +1,108 @@
+#include <faabric/transport/MessageEndpointServer.h>
+
+// Defined in libzmq/include/zmq.h (156384765)
+#define ZMQ_ETERM ETERM
+
+namespace faabric::transport {
+MessageEndpointServer::MessageEndpointServer(const std::string& hostIn,
+                                             int portIn)
+  : host(hostIn)
+  , port(portIn)
+{}
+
+void MessageEndpointServer::start()
+{
+    start(faabric::transport::getGlobalMessageContext());
+}
+
+void MessageEndpointServer::start(faabric::transport::MessageContext& context)
+{
+    // Start serving thread in background
+    this->servingThread = std::thread([this, &context] {
+        MessageEndpointClient serverEndpoint(this->host, this->port);
+
+        // Open message endpoint, and bind
+        serverEndpoint.open(
+          context, faabric::transport::SocketType::PULL, true);
+
+        // Loop until context is terminated
+        while (true) {
+            try {
+                this->recv(serverEndpoint);
+            } catch (zmq::error_t& e) {
+                if (e.num() == ZMQ_ETERM) {
+                    serverEndpoint.close();
+                    break;
+                }
+                throw std::runtime_error(fmt::format(
+                  "Errror in socket receiving message: {}", e.what()));
+            }
+        }
+    });
+}
+
+void MessageEndpointServer::stop()
+{
+    stop(faabric::transport::getGlobalMessageContext());
+}
+
+void MessageEndpointServer::stop(faabric::transport::MessageContext& context)
+{
+    // Note - different servers will concurrently close the server context, but
+    // this structure is thread-safe, and the close operation idempotent.
+    context.close();
+
+    // Finally join the serving thread
+    if (this->servingThread.joinable()) {
+        this->servingThread.join();
+    }
+}
+
+void MessageEndpointServer::recv(MessageEndpointClient& endpoint)
+{
+    // Receive header
+    if (!endpoint.socket) {
```

Comment:
Generally I would avoid relying on an object's `operator !` if you are just checking for `nullptr`, and instead write `if(endpoint.socket == nullptr)`. As it stands, this condition could evaluate to `false` when the socket is _not_ a null pointer (depending on the implementation of `operator !`), thus leading to an invalid error message in this case.